godot-core/src/obj/script.rs

```diff
@@ -337,6 +337,33 @@ pub unsafe fn create_script_instance<T: ScriptInstance>(
     }
 }
 
+/// Checks if a script instance exists for a given object.
+///
+/// The engine does not expose the [`ScriptLanguage`] of a [`Script`] (as of 4.3) and it's therefore necessary to pass the expected
+/// language as well.
+///
+/// Use this function to implement [`IScriptExtension::instance_has`](crate::classes::IScriptExtension::instance_has).
+#[cfg(since_api = "4.2")]
+pub fn script_instance_exists<O, S, L>(object: &Gd<O>, script: &Gd<S>, language: &Gd<L>) -> bool
+where
+    O: Inherits<Object>,
+    S: Inherits<Script>,
+    L: Inherits<ScriptLanguage>,
+{
+    let get_instance_fn = sys::interface_fn!(object_get_script_instance);
+
+    let object_script_variant = object.upcast_ref().get_script();
+
+    if object_script_variant.is_nil() {
+        return false;
+    }
+
+    let object_script: Gd<Script> = object_script_variant.to();
+    let instance = unsafe { get_instance_fn(object.obj_sys(), language.obj_sys()) };
+
+    !instance.is_null() && object_script == script.clone().upcast()
```

Comment:
Since 4.4, you can use `Variant::object_id()`:

https://github.com/godot-rust/gdext/blob/40186f3a5afebafb7b86ce06a7d676bf418c646e/godot-core/src/builtin/variant/mod.rs#L110-L122

Since you only want to check for equality, this would save the `Variant::to()` marshalling, the `clone()` and the `upcast()` conversion.

Maybe something like:
```suggestion
    let instance = unsafe { get_instance_fn(object.obj_sys(), language.obj_sys()) };
    
    let script_id = object_script_variant.object_id().expect(...);
    !instance.is_null() && object_script.instance_id() == script_id
```


I tried backporting `Variant::object_id()` to <4.4 so you can use it universally, but encountered a bug while doing so. Will follow up on this :+1:

Reply:
So I was able to use `upcast_ref()` and `__object_id()` for the comparison, but the `Variant::to()` marshalling without `Variant::object_id()` right now.